### PR TITLE
Data monitor

### PR DESCRIPTION
**Pull request by jhonabreul**
#### Description

New `DataMonitor` class to keep track of data requests statistics:
- Total and successful/failed requests
- Universe data requests
- Request rates
- Report generation with summary of these stats.

#### Related Issue

N/A

#### Motivation and Context

N/A

#### Requires Documentation Change

N/A

#### How Has This Been Tested?

Unit tests, backtests.

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Common/DataMonitor.cs [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+        private Thread _requestRateCalculationThread;
+        private CancellationTokenSource _cancellationTokenSource;
+
+        private string _succeededDataRequestsFileName;
+        private string _failedDataRequestsFileName;
+
+        /// <summary>
+        /// Directory location to store results
+        /// </summary>
+        public string ResultsDestinationFolder { get; }
```
Comment (Martin-Molinero) [inline Common/DataMonitor.cs:56]: Could be a private field? Seems we don't need a public property

File: Common/DataMonitor.cs [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+        /// <param name="report">The data monitor report to be stored<param>
+        private void StoreDataMonitorReport(DataMonitorReport report)
+        {
+            if (report == null)
+            {
+                return;
+            }
+
+            var path = GetFilePath("data-monitor-report.json");
+            var data = JsonConvert.SerializeObject(report, Formatting.None  );
```
Comment (Martin-Molinero) [inline Common/DataMonitor.cs:238]: `Formatting.None  );` nit spaces

`GetFilePath` is adding the timestamp after the file extension, ideally it's before it

File: Common/DataMonitor.cs [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+        private void Initialize()
+        {
+            if (_requestRateCalculationThread != null)
+            {
+                return;
+            }
+            
+            _cancellationTokenSource = new CancellationTokenSource();
+
+            _requestRateCalculationThread = new Thread(() =>
```
Comment (Martin-Molinero) [inline Common/DataMonitor.cs:172]: Cool! we should take a lock here to make sure we start a single one, after the initial null check, there could be multiple threads fetching files at the same time 👍

**Review by Martin-Molinero — COMMENTED** [review]
Nice! Leaving couple of minor requests 👍

File: Common/DataMonitor.cs (hunk 2 above)
Comment (jhonabreul) [reply Common/DataMonitor.cs:238]: My  bad with the timestamp, it has to be before the tghe extension 🎯

**Review by Martin-Molinero — APPROVED** [review]

**Pull request merged**